### DISCUSSION TITLE
Do you know of a way to see a list of users that are not associated with any group?

### DISCUSSION THREAD

**Discussion by bmofkn** [thoughtspot/cs_tools#113] [Q&A] [answered]
I've searched the TS Community forums and found [this post](https://community.thoughtspot.com/customers/s/question/0D53n00008Y0edUCAR/do-you-know-of-a-way-to-see-a-list-of-users-that-are-not-associated-with-any-group?t=1711375566110).  It references cs_tools but the links in it are outdated.

Does anyone have an example cs_tools script to return all users that do not belong to a group? I know we can try something like the following... 

`cs_tools tools searchable gather --config prod_ts --syncer csv:///Users/bryan/definition.toml`

... but that returns 12 CSV files (of which three are ts_user, ts_group and ts_xref_principal).  I'd just like to get a list of users that don't belong to a group.  Are there any parameters we can pass to slim the response down to just users without a group?

**Comment by boonhapus** [comment, marked as answer]
Hey @bmofkn , if you sync to SQLite then you can write the following query against the data. **[I personally use DBeaver](https://dbeaver.io/)** as it's free.

`cs_tools tools searchable gather --config prod_ts --syncer sqlite://database_path=prod.db`

```sql
SELECT
      u.username
    , u.display_name
FROM
    TS_USER AS u
LEFT JOIN
    TS_XREF_PRINCIPAL AS m ON m.principal_guid = u.user_guid
WHERE
    m.principal_guid IS NULL  -- corrected to xref, instead of user... duh 😅
```

**Reply by boonhapus** [reply]
@bmofkn this might be a little bit more plug and play in case you're on TS Cloud which has a few default groups.

```sql
SELECT
      u.username
    , u.display_name
FROM
    TS_USER AS u
LEFT JOIN
    TS_XREF_PRINCIPAL AS m ON m.principal_guid = u.user_guid
LEFT JOIN
    TS_GROUP AS g ON g.group_guid = m.group_guid
WHERE
    g.group_name NOT IN ('Demo Retail Group', 'All Orgs', 'System Management Group')
GROUP BY
    u.username, u.display_name
HAVING
    COUNT(g.group_name) = 0 
ORDER BY
    u.username
```

**Reply by bmofkn** [reply]
Thanks @boonhapus, wow, using sqlite is so much better than CSV!  

Unfortunately, it looks like every user belongs to a single phantom group... perhaps that's what you were trying to allude to with your revised SQL.  But the phantom `group_guid` doesn't appear in `TS_GROUP`... it only shows up in `TS_XREF_PRINCIPAL`.  I tried using `SearchUserGroups` in GraphQL to get details on this phantom group but it always returns Null.  I do see 'All Orgs', 'System Management Group' and 'Demo Retail Group', none of which are the phantom group 🤔

However, if I filter out the phantom `group_guid`, I get the list I'm seeking 🙂

Thanks again for the quick replies.  Cheers!

**Reply by boonhapus** [reply]
This might be the `All Group` , which used to be named but seems to be hidden now so it can't be fetched via the Group API in the extract. This is likely a bug in ThoughtSpot APIs... I'll investigate it a bit internally and see what's up.

**Reply by AlexanderStephenson** [reply]
I believe this is the All group that all users belong to.  I have asked in Product Ideas that this group be exposed but not editable.  I believe this group is used for the Share to All users privilege.

You can use this to your advantage, as you know all users must belong to this group and you do not need left outer joins, and can filter on users that only belong to one group.

I am processing the CS Tools tables via dbt to create some subtle changes to the tables, and clearly mark what I think is a DIM and a FACT.  For my dim__groups I add in missing entries for the groups that are in the XREF table that do not exist in the Groups table.

`with

int_ts__group as 
(
    select * from {{ ref('int_ts__group') }}
),

int_ts__xref_principal as 
(
    select * from {{ ref('int_ts__xref_principal') }}
),

missing_xref_groups as
(
    select distinct
        int_ts__xref_principal.group_guid as group_guid,
        'missing object' as group_name,
        'missing object' as description,
        'missing object' as display_name,
        'missing object' as sharing_visibility,
        null as created,
        null as modified,
        'unknown xref group' as group_type

    from int_ts__xref_principal
    left join int_ts__group on
        int_ts__xref_principal.group_guid = int_ts__group.group_guid

    where int_ts__group.group_guid is null
),

dim_ts__group as 
(

    select
        group_guid,
        group_name,
        description,
        display_name,
        sharing_visibility,
        created,
        modified,
        group_type

    from int_ts__group

    union all

    select
        group_guid,
        group_name,
        description,
        display_name,
        sharing_visibility,
        created,
        modified,
        group_type

    from missing_xref_groups

)

select * from dim_ts__group
order by group_guid`

I have created a worksheet in ThoughtSpot and from this the SQL generated is as follows.  This has highlighted 57 users that do not belong to any group.  So quite a handy idea as I can delete these user accounts as they have no access to anything.
`SELECT 
  "ta_1"."USERNAME" "ca_1", 
  (NOT("ta_2"."PRINCIPAL_GUID" IS NULL)) "ca_2"
FROM "ANALYTICS_DEV"."DBT_ASTEPHENSON"."FCT_TS__XREF_PRINCIPAL" "ta_2"
  JOIN "ANALYTICS_DEV"."DBT_ASTEPHENSON"."DIM_TS__USER" "ta_1"
    ON "ta_2"."PRINCIPAL_GUID" = "ta_1"."USER_GUID"
  JOIN "ANALYTICS_DEV"."DBT_ASTEPHENSON"."DIM_TS__GROUP" "ta_3"
    ON "ta_2"."GROUP_GUID" = "ta_3"."GROUP_GUID"
GROUP BY 
  "ca_1", 
  "ca_2"
HAVING count(DISTINCT "ta_3"."GROUP_NAME") = 1`
